### ISSUE TITLE
Move support libraries to androidx

### ISSUE THREAD

**Issue by mcomella** [mozilla-mobile/firefox-tv#1398]
### Why/User Benefit/User Problem
Like #991, we have to stay up to date with the latest android conventions. Staying up to date ensures we can add new android features for users easily. This migration is part of that. I don't know if there's a date that we have to do this by but I'd assume it's by next fall, for SDK 29. That being said, components and other libraries that migrate may force us to adopt this so we should do it sooner.

### What / Requirements
- Use migration tool to move to new support libraries
- Use jetifier to update dependencies to use androidx too
- Add new toys that we're unable to access like androidx KTX extensions (see [https://developer.android.com/topic/libraries/architecture/adding-components](https://developer.android.com/topic/libraries/architecture/adding-components)

### Acceptance Criteria (how do I know when I’m done?)
- QA acceptance that there are no regressions
  - Focus on dependency integrations (b/c jetifier rewrites)

**Comment by psymoon** [comment]
Required for https://github.com/mozilla-mobile/firefox-tv/issues/1391

See: http://robolectric.org/migrating/

**Comment by pocmo** [comment]
FYI: I copied this issue to our large AndroidX migration meta:
https://github.com/mozilla-mobile/android-components/issues/1903

TL;DR: We need downstream app projects to migrate first before we can migrate up the dependency tree.

Is this something we could contribute? Is there a good time in your release schedule where such a migration would fit (and get the required QA attention)?

**Comment by mcomella** [comment]
@pocmo  It looks like this issue has not been triaged yet. If we were to tackle this, we'd prioritize it in triage next Wednesday: can we get back to you then?

> Is this something we could contribute?

Yeah! I don't see why not. Let us know if you decide to take this route.

...good luck! 😝 

> Is there a good time in your release schedule where such a migration would fit (and get the required QA attention)?

I can't speak for the team and as the FFES dev, I'm missing context, but we are just finishing with the regression whack-a-mole from our refactor so I'd think this could be something to land once the refactor is stable (after maybe 1 or 2 releases?). Last week's release is was backed out – it's now in staged rollout – and we're working on the following release currently.

**Comment by pocmo** [comment]
> @pocmo It looks like this issue has not been triaged yet. If we were to tackle this, we'd prioritize it in triage next Wednesday: can we get back to you then?

Sure! :)

**Comment by Baron-Severin** [comment]
Some notes as I'm making these changes.

- The Google provided migration script is incomplete (`android.support.v4.media.session.PlaybackStateCompat` is not updated, should move to `android.media.session.PlaybackState`)
- We are already on support lib v28.0.0, which helps make this move easier
- Our Gradle plugin needs to be updated to at least 3.2.0
- Jetifier (updates dependencies that rely on support libs to use AndroidX equivelents) is not working for some dependencies. Not sure why yet (TODO will update here with a complete list)

**Comment by Baron-Severin** [comment]
Still no idea why, but the following dependencies:
```
org.mozilla.components:browser-session:$moz_components_version
org.mozilla.components:browser-search:$moz_components_version
org.mozilla.components:concept-engine:$moz_components_version
org.mozilla.components:service-telemetry:$moz_components_version
org.mozilla.components:feature-session:$moz_components_version
org.mozilla.components:service-fretboard:$moz_components_version
org.mozilla.components:browser-engine-system:$moz_components_version
```
all fail unless they declare `exclude group: 'com.android.support'`.  The resulting error when syncing is:
```
Manifest merger failed : Attribute application@appComponentFactory value=(androidx.core.app.CoreComponentFactory) from [androidx.core:core:1.0.0] AndroidManifest.xml:22:18-86
	is also present at [com.android.support:support-compat:28.0.0] AndroidManifest.xml:22:18-91 value=(android.support.v4.app.CoreComponentFactory).
	Suggestion: add 'tools:replace="android:appComponentFactory"' to <application> element at AndroidManifest.xml:22:5-122:19 to override.
```

**Comment by Baron-Severin** [comment]
Updated notes on this migration.
* VideoVoiceCommandMediaSession does not update correctly
    * import android.support.v4.media.session.PlaybackStateCompat.ACTION_PAUSE -> import android.media.session.PlaybackState.ACTION_PAUSE
* Jetifier doesn't seem to work for some dependencies.  Why not? (SOLVED, see below)
    * This was because we had our Gradle version declared in two places, and one had not been updated
        * Jetifier was not running at all
        * The below dependencies were just the ones that immediately blew up as a result
    * Affected dependencies
        * org.mozilla.components:browser-session:$moz_components_version
        * org.mozilla.components:browser-search:$moz_components_version
        * org.mozilla.components:concept-engine:$moz_components_version
        * org.mozilla.components:service-telemetry:$moz_components_version
        * org.mozilla.components:feature-session:$moz_components_version
        * org.mozilla.components:service-fretboard:$moz_components_version
        * org.mozilla.components:browser-engine-system:$moz_components_version
    * Exclude seems to fix Fastlane builds (but screenshots may or may not work?)
    * Moz components can't exclude.  We depend on LifecycleOwner in a lot of cases
        * We can manually Jetify using https://developer.android.com/studio/command-line/jetifier
        * Why doesn't this work automatically?
* Gradle plugin needs to be updated to at least 3.2.0
    * 3.3.1 seems fine
* Some dependencies (Fastlane screenGrab) depend upon older versions of support libraries
    * These can't be Jetified
    * Only solution I've found is to exclude those to prevent conflicts
* We had problems with ApplicationProvider
    * Dependency was downloaded, Intellij recognized it, but at build time it blew up with an unrecognized reference error
        * Note the order of the above, this is NOT "compiler recognized, Intellij did not", it is the reverse
        * I was unable to solve this
        * Substituted `InstrumentationRegistry.getInstrumentation().context.applicationContext` for `ApplicationProvider.getApplicationContext()`
* `./gradlew testSystemDebug --info` fails with the message "Transformation hasn't been executed yet"
    * Still working on this
